service/topology_coordinator.cc

```diff
@@ -1598,6 +1600,20 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
             co_return;
         }
 
+        // generate updates for pending resize finalizations
+        if (has_pending_resize_finalization && !utils::get_local_injector().enter("tablet_split_finalization_within_tablet_migration_postpone")) {
+            updates.clear();
+            guard = co_await maybe_generate_resize_finalization_updates(updates, std::move(guard));
+
+            // `maybe_generate_resize_finalization_updates` will generate finalization updates iff there are no active transitions.
+            // So, if updates are not empty, we can continue or else, we need to return to handle the active transitions.
+            if (updates.empty()) {
+                // transitions were started when `maybe_generate_resize_finalization_updates` attempted
+                // to take global barrier - cannot continue with resize finalizations.
+                co_return;
```

Comment:
So if there were finalizing updates, we will try to exit the "tablet_migration" or "tablet_draining" transition. That's not good. If we're draining, we will hit the `if (has_nodes_to_drain)` case and not execute the updates, and will be stuck forever. If we're not draining, it's also not nice because after finalization there could be load-balancing migrations to perform, and we will have this period with null tstate before we enter "tablet_migration" again, and that can cause await_topology_quiesced() to exit prematurely as soon as it observes null tstate.

We should rather execute the finalizing updates `if (updates)` here. And co_return in any case.